sql: add relevant udfs to statement bundles

lyang24 opened the pull request:
Statement bundles now include the CREATE statements for relevant udf in schema.sql.

Epic: None
Fixes: https://github.com/cockroachdb/cockroach/issues/102044

Release note: None

Commit:
Statement bundles now include the CREATE statements for related udfs in schema.sql.

Epic: None
Fixes: #102044

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/102860)
<!-- Reviewable:end -->

rytaft reviewed (commented):
Thank you for the contribution, @lyang24, this is great!

I'm adding @michae2 as a reviewer since he recently worked on statement bundles.

> Reviewed 2 of 2 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/102860#-:-NV6gCKJ7R_QCGaLepta:b-cjvu76)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @lyang24 and @michae2)

___
*[`pkg/sql/explain_bundle.go` line 826 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/102860#-NV6gSCV95xLLXIB0751:-NV6gSCV95xLLXIB0752:bo9pcj2) ([raw file](https://github.com/cockroachdb/cockroach/blob/18306022523b3991fc11cdade748a4dd7291ca06/pkg/sql/explain_bundle.go#L826)):*
> ```Go
> 
> func (c *stmtEnvCollector) PrintRelevantCreateUdf(w io.Writer, stmt string) error {
> 	qry := "SELECT function_name FROM crdb_internal.create_function_statements;"
> ```

I think we'll need to add a version of this that redacts constants if `b.flags.RedactValues` is true, similar  to the `PrintCreateEnum` function above.


<!-- Sent from Reviewable.io -->

michae2 reviewed (commented):
Thank you for doing this!

> Reviewed 2 of 2 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/102860#-:-NV73CcTDUUcqrjiYYTx:bff479j)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @lyang24 and @rytaft)

___
*[`pkg/sql/explain_bundle.go` line 826 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/102860#-NV71byG5qBeA-aQsXZX:-NV71byG5qBeA-aQsXZY:b2sq2ue) ([raw file](https://github.com/cockroachdb/cockroach/blob/18306022523b3991fc11cdade748a4dd7291ca06/pkg/sql/explain_bundle.go#L826)):*
> ```Go
> 
> func (c *stmtEnvCollector) PrintRelevantCreateUdf(w io.Writer, stmt string) error {
> 	qry := "SELECT function_name FROM crdb_internal.create_function_statements;"
> ```

nit: I think it would be a little more future-proof to go through `SHOW FUNCTIONS` rather than `crdb_internal.create_function_statements`, since `crdb_internal` tables can change without warning. I suggest something like: `SELECT lower(function_name) FROM [SHOW FUNCTIONS]`.
___
*[`pkg/sql/explain_bundle.go` line 826 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/102860#-NV6gSCV95xLLXIB0751:-NV73FPQARcWliqVgIyw:b32v57) ([raw file](https://github.com/cockroachdb/cockroach/blob/18306022523b3991fc11cdade748a4dd7291ca06/pkg/sql/explain_bundle.go#L826)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

I think we'll need to add a version of this that redacts constants if `b.flags.RedactValues` is true, similar  to the `PrintCreateEnum` function above.
</blockquote></details>

+1
___
*[`pkg/sql/explain_bundle.go` line 834 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/102860#-NV72oXC3Q5RHM1bOs4A:-NV72oXC3Q5RHM1bOs4B:b-t171tz) ([raw file](https://github.com/cockroachdb/cockroach/blob/18306022523b3991fc11cdade748a4dd7291ca06/pkg/sql/explain_bundle.go#L834)):*
> ```Go
> 		if strings.Contains(stmt, name) {
> 			createStatement, err := c.query(fmt.Sprintf(
> 				"SELECT create_statement FROM crdb_internal.create_function_statements WHERE function_name = '%s'", name,
> ```

nit: Same comment here, I would go through `SHOW CREATE FUNCTION` to make it a little more future-proof.


<!-- Sent from Reviewable.io -->

lyang24 commented:
> Thank you for doing this!
> 
> > Reviewed 2 of 2 files at r1, all commit messages.
> > _[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/102860#-:-NV73CcTDUUcqrjiYYTx:bff479j)_ status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @lyang24 and @rytaft)
> 
> _[`pkg/sql/explain_bundle.go` line 826 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/102860#-NV71byG5qBeA-aQsXZX:-NV71byG5qBeA-aQsXZY:b2sq2ue) ([raw file](https://github.com/cockroachdb/cockroach/blob/18306022523b3991fc11cdade748a4dd7291ca06/pkg/sql/explain_bundle.go#L826)):_
> 
> > ```go
> > func (c *stmtEnvCollector) PrintRelevantCreateUdf(w io.Writer, stmt string) error {
> > 	qry := "SELECT function_name FROM crdb_internal.create_function_statements;"
> > ```
> 
> nit: I think it would be a little more future-proof to go through `SHOW FUNCTIONS` rather than `crdb_internal.create_function_statements`, since `crdb_internal` tables can change without warning. I suggest something like: `SELECT lower(function_name) FROM [SHOW FUNCTIONS]`.
> 
> _[`pkg/sql/explain_bundle.go` line 826 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/102860#-NV6gSCV95xLLXIB0751:-NV73FPQARcWliqVgIyw:b32v57) ([raw file](https://github.com/cockroachdb/cockroach/blob/18306022523b3991fc11cdade748a4dd7291ca06/pkg/sql/explain_bundle.go#L826)):_
> 
> Previously, rytaft (Rebecca Taft) wrote…
> +1
> 
> _[`pkg/sql/explain_bundle.go` line 834 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/102860#-NV72oXC3Q5RHM1bOs4A:-NV72oXC3Q5RHM1bOs4B:b-t171tz) ([raw file](https://github.com/cockroachdb/cockroach/blob/18306022523b3991fc11cdade748a4dd7291ca06/pkg/sql/explain_bundle.go#L834)):_
> 
> > ```go
> > 		if strings.Contains(stmt, name) {
> > 			createStatement, err := c.query(fmt.Sprintf(
> > 				"SELECT create_statement FROM crdb_internal.create_function_statements WHERE function_name = '%s'", name,
> > ```
> 
> nit: Same comment here, I would go through `SHOW CREATE FUNCTION` to make it a little more future-proof.

Thank you for the feedback. I have addressed the comments and fixed a couple of nits. Looking forward for a second look at this pr.

rytaft reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>, just a couple of minor nits. Thanks again for the contribution!

@michae2, will the redaction automatically be tested by your randomized tests, or do we need to do something to test it?

> Reviewed 2 of 2 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/102860#-:-NVUmqaO-NDvaRUTEPCx:bf1p9jr)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @lyang24 and @michae2)

___
*[`pkg/sql/explain_bundle.go` line 830 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/102860#-NVUlrvf-A4m1HDqDMk5:-NVUlrvf-A4m1HDqDMk6:b-waptg0) ([raw file](https://github.com/cockroachdb/cockroach/blob/053b6cb3ecbcdee75dba3e1e13b6326a5437fb47/pkg/sql/explain_bundle.go#L830)):*
> ```Go
> 	w io.Writer, stmt string, redactValues bool,
> ) error {
> 	// The select function_name returns a DOidWrapper, we need to cast it to string for queryRows function to process.
> ```

nit: comments should be <= 80 characters wide.
___
*[`pkg/sql/opt/metadata.go` line 543 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/102860#-NVUmNvPB-yQgPDOvoxf:-NVUmNvPB-yQgPDOvoxg:b9ghs31) ([raw file](https://github.com/cockroachdb/cockroach/blob/053b6cb3ecbcdee75dba3e1e13b6326a5437fb47/pkg/sql/opt/metadata.go#L543)):*
> ```Go
> }
> 
> // AllUserDefinedFunctions returns all user defined functions related in this query.
> ```

nit: "related in" -> "used in"


<!-- Sent from Reviewable.io -->

michae2 reviewed (changes requested):
> Reviewed 2 of 2 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/102860)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @lyang24)

___
*[`pkg/sql/explain_bundle.go` line 849 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/102860#-NVVhBeQ5h5kQUqvav_Y:-NVVhBeQ5h5kQUqvav_Z:b-lem0zi) ([raw file](https://github.com/cockroachdb/cockroach/blob/053b6cb3ecbcdee75dba3e1e13b6326a5437fb47/pkg/sql/explain_bundle.go#L849)):*
> ```Go
> 			createStatement, err := c.query(createFunctionQuery)
> 			if err != nil {
> 				return err
> ```

Even if we fail to get the definition for one UDF, we might have successfully gotten the definition of other UDFs used by the query. So instead of returning the error here, I propose we print some kind of message with the name of the UDF we failed on (would be nice if the message started with SQL comments `--` so that we can still execute the file) and `continue` in the loop.


<!-- Sent from Reviewable.io -->

lyang24 commented:
> > Reviewed 2 of 2 files at r2, all commit messages.
> > _[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/102860)_ status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @lyang24)
> 
> _[`pkg/sql/explain_bundle.go` line 849 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/102860#-NVVhBeQ5h5kQUqvav_Y:-NVVhBeQ5h5kQUqvav_Z:b-lem0zi) ([raw file](https://github.com/cockroachdb/cockroach/blob/053b6cb3ecbcdee75dba3e1e13b6326a5437fb47/pkg/sql/explain_bundle.go#L849)):_
> 
> > ```go
> > 			createStatement, err := c.query(createFunctionQuery)
> > 			if err != nil {
> > 				return err
> > ```
> 
> Even if we fail to get the definition for one UDF, we might have successfully gotten the definition of other UDFs used by the query. So instead of returning the error here, I propose we print some kind of message with the name of the UDF we failed on (would be nice if the message started with SQL comments `--` so that we can still execute the file) and `continue` in the loop.

great catch i fixed it.

rytaft reviewed (approved):
> Reviewed 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/102860)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @lyang24 and @michae2)

___
*[`pkg/sql/explain_bundle.go` line 830 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/102860#-NVUlrvf-A4m1HDqDMk5:-NVWJN1X197wcockGleI:b2wdavv) ([raw file](https://github.com/cockroachdb/cockroach/blob/053b6cb3ecbcdee75dba3e1e13b6326a5437fb47/pkg/sql/explain_bundle.go#L830)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

nit: comments should be <= 80 characters wide.
</blockquote></details>

Thanks, although I thought the original comment was good, just wrap the lines


<!-- Sent from Reviewable.io -->

michae2 reviewed (changes requested):
@rytaft good catch. The randomized tests check redaction of most forms of `EXPLAIN` and `EXPLAIN ANALYZE` but not `EXPLAIN ANALYZE (DEBUG)`. Only `TestExplainAnalyzeDebug/redact` checks that. @lyang24 I left a comment below about how you could change this test to also check redaction of UDF definitions.

> Reviewed 2 of 2 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/102860#-:-NVWL1Ff2l5yVI5n31BM:b878ie9)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @lyang24 and @rytaft)

___
*[`pkg/sql/explain_bundle_test.go` line 292 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/102860#-NVWLetPCYbKcSHZauGe:-NVWLetPCYbKcSHZauGf:b-tdbs4l) ([raw file](https://github.com/cockroachdb/cockroach/blob/15880537de1a0cb2b682d55487ba4e69f50ea0ff/pkg/sql/explain_bundle_test.go#L292)):*
> ```Go
> 		r.Exec(t, "CREATE STATISTICS jurassic FROM pterosaur;")
> 		rows := r.QueryStr(t,
> 			"EXPLAIN ANALYZE (DEBUG, REDACT) SELECT max(cardno) FROM pterosaur WHERE cardholder = 'pterodactyl'",
> ```

To test redaction, you can add a UDF with the string constant `'pterodactyl'` in its body, and call it from this `SELECT` to make sure that the UDF definition is redacted in the bundle.


<!-- Sent from Reviewable.io -->

lyang24 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/102860)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @michae2 and @rytaft)

___
*[`pkg/sql/explain_bundle.go` line 830 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/102860#-NVUlrvf-A4m1HDqDMk5:-NVWMx_F7DH4ZSCWHWtj:b9aextj) ([raw file](https://github.com/cockroachdb/cockroach/blob/053b6cb3ecbcdee75dba3e1e13b6326a5437fb47/pkg/sql/explain_bundle.go#L830)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Thanks, although I thought the original comment was good, just wrap the lines
</blockquote></details>

I see, reverted and added wrapping.
___
*[`pkg/sql/explain_bundle_test.go` line 292 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/102860#-NVWLetPCYbKcSHZauGe:-NVWRrk_1vOrIFpZimEd:bjav0yp) ([raw file](https://github.com/cockroachdb/cockroach/blob/15880537de1a0cb2b682d55487ba4e69f50ea0ff/pkg/sql/explain_bundle_test.go#L292)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

To test redaction, you can add a UDF with the string constant `'pterodactyl'` in its body, and call it from this `SELECT` to make sure that the UDF definition is redacted in the bundle.
</blockquote></details>

thank you I modified TestExplainAnalyzeDebug/redact to include the suggested change


<!-- Sent from Reviewable.io -->

rytaft reviewed (commented):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> please go ahead and merge @michae2 if it looks good to you. Thanks @lyang24!

> Reviewed 1 of 6 files at r4, 1 of 6 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/102860#-:-NVWjr4m8zUMu2tdFYbk:b-qz87om)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @michae2)



<!-- Sent from Reviewable.io -->

michae2 reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Thank you @lyang24! I'll merge this.

bors r=rytaft,michae2 a=lyang24

> Reviewed 1 of 6 files at r4, 1 of 6 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/102860#-:-NVWvir3D7TDyhx7yc1Q:b-hs5z5z)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @rytaft)



<!-- Sent from Reviewable.io -->
